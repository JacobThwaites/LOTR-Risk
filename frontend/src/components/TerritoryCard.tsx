import React from "react";
import eagle from '../assets/eagle.svg';
import nazgul from '../assets/nazgul.png';
import archer from '../assets/archer.svg';
import wildCard from '../assets/wild-card.png';

export default function TerritoryCard(props: { card: string, index: number, isSelected: boolean, onClick: any }): JSX.Element {
    let className = 'territory-cards--card';
  
    if (props.isSelected) {
      className += ' selected';
    }
  
    function getImageSource(symbol: string): string {
      if (symbol === "ARCHER") {
        return archer;
      } else if (symbol === "CAVALRY") {
        return nazgul;
      } else if (symbol === "EAGLE") {
        return eagle;
      } else {
        return wildCard;
      }
    }
  
    return (
      <div className={className} key={props.index} onClick={props.onClick} data-testid={className}>
        <div className='territory-card--card_symbol'>
          <img className={`symbol-${props.card}`} src={getImageSource(props.card)} alt={props.card}/>
        </div>
        <p>
          {props.card}
        </p>
      </div>
    );
  }