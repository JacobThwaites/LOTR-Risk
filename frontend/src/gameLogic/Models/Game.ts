import { Player } from './Player';

export class Game {
    private players: Array<Player>;
    private maxTurns: number;
    private currentTurn: number;
    private currentPlayersTurn: number;
    constructor(players: Array<Player>, maxTurns: number) {
        this.players = players;
        this.maxTurns = maxTurns;
        this.currentTurn = 0;
        this.currentPlayersTurn = 0;
    }

    getPlayers(): Array<Player> {
        return this.players;
    }

    getCurrentPlayer(): Player {
        const indexOfPlayer = this.currentPlayersTurn;
        return this.players[indexOfPlayer];
    }

    getTurnsRemaining(): number {
        return this.maxTurns - this.currentTurn;
    }

    handleNewTurn() {
        this.changeCurrentPlayer();
        const newCurrentPlayer = this.getCurrentPlayer();
        newCurrentPlayer.addReinforcementsForNewTurn();
    }

    changeCurrentPlayer() {
        this.currentPlayersTurn += 1;
        const lastPlayer = this.players.length - 1;
        if (this.currentPlayersTurn > lastPlayer) {
            this.currentPlayersTurn = 0;
            this.incrementCurrentTurn();
        }
    }

    assignStartingUnits() {
        const unitsAvailable = this.getStartingUnitsAvailable();

        for (let i = 0; i < this.players.length; i++) {
            this.players[i].addReinforcements(unitsAvailable);
            this.players[i].addStartingUnits();
        }
    }

    getStartingUnitsAvailable(): number {
        // TODO: remove after testing
        return 34;
        // if (this.players.length === 2) {
        //     return 60;
        // } else if (this.players.length === 3) {
        //     return 52;
        // } else {
        //     return 45;
        // }
    }

    playersHaveReinforcements(): boolean {
        for (let i = 0; i < this.players.length; i++) {
            if (this.players[i].getReinforcements() > 0) {
                return true;
            }
        }

        return false;
    }

    incrementCurrentTurn() {
        this.currentTurn++;
    }

    checkMaxTurnsReached(): boolean {
        return this.maxTurns <= this.currentTurn;
    }

    getNextUnusedPlayer(): Player | boolean {
        for (let i = 0; i < this.players.length; i++) {
            if (!this.players[i].getUserID()) {
                return this.players[i];
            }
        }

        return false;
    }
}