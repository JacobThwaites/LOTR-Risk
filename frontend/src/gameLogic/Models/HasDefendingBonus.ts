import { IDefendingBonus } from "./IDefendingBonus";

export class HasDefendingBonus implements IDefendingBonus {
    public getBonus(): number {
        return 1;
    }
}