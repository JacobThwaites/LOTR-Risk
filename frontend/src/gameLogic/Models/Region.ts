import { AreaType } from './AreaType';

export class Region {
    private name: string;
    private areas: Array<AreaType>;
    private bonusUnits: number;
    constructor(name: string, areas: Array<AreaType>, bonusUnits: number) {
        this.name = name;
        this.areas = areas;
        this.bonusUnits = bonusUnits;
    }

    getName(): string {
        return this.name;
    }

    getAreas(): Array<AreaType> {
        return this.areas;
    }

    getBonusUnits(): number {
        return this.bonusUnits;
    }
}