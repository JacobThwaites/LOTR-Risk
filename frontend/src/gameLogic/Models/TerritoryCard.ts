import { Symbol } from '../Enums/Symbols';
import { AreaType } from './AreaType';

export class TerritoryCard {
    private area: AreaType;
    private symbol: Symbol;
    constructor(area: AreaType, symbol: Symbol) {
        this.area = area;
        this.symbol = symbol;
    }

    getArea(): AreaType {
        return this.area;
    }

    getSymbolValue(): Symbol {
        return this.symbol;
    }
}