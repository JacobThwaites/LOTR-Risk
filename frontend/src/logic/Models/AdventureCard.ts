export class AdventureCard {
    //TODO: amend effect to be an interface
    private effect: string;
    constructor(effect: string) {
        this.effect = effect;
    }

    getEffect(): string {
        return this.effect;
    }
}