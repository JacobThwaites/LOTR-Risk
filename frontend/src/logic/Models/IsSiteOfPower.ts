import { ISiteOfPower } from "./ISiteOfPower";

export class IsSiteOfPower implements ISiteOfPower {
    public isSiteOfPower(): boolean {
        return true;
    }
}