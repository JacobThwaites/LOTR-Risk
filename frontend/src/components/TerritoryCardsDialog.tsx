import React, { useState } from "react";
import { Dialog, DialogContent } from '@mui/material';
import { TerritoryCard } from "../gameLogic/Models/TerritoryCard";
import TerritoryCardManager, { TradableCards } from "../gameLogic/Controllers/TerritoryCardManager";
import { Symbol } from "../gameLogic/Enums/Symbols";
import eagle from '../assets/eagle.svg';
import nazgul from '../assets/nazgul.png';
import archer from '../assets/archer.svg';
import wildCard from '../assets/wild-card.png';
import CustomButton from "./common/CustomButton";


type Props = { 
  onClose: any, 
  cards: TerritoryCard[]
}

export default function TerritoryCardsDialog(props: Props): JSX.Element {
  const [selectedCards, setSelectedCards] = useState<TerritoryCard[]>([]);

  function onCardSelect(card: TerritoryCard): void {
    if (selectedCards.includes(card)) {
      const newCards = selectedCards.filter(c => c !== card);
      setSelectedCards(newCards);
    } else if (TerritoryCardManager.isNewCardValidWithExistingCards(selectedCards, card)) {
      setSelectedCards([...selectedCards, card]);
    }
  }

  function areSelectedCardsExchangeable(): boolean {
    return selectedCards.length === 3 && TerritoryCardManager.areCardsExchangeable(selectedCards as TradableCards);
  }

  function tradeCards(): void {
    // TerritoryCardManager.exchangeCards(props.player, selectedCards as TradableCards);

    // TODO: send websocket message
    // props.updateGameState();
    props.onClose();
  }

  const cards = props.cards.map((card, i) => {
    return <Card key={i} card={card} index={i} isSelected={selectedCards.includes(card)} onClick={() => onCardSelect(card)} />
  })

  return (
    <Dialog onClose={props.onClose} open={true} fullWidth maxWidth='sm' className='territory-cards' scroll='paper'>
      <h1 className='territory-cards--header'>Territory Cards</h1>
      <DialogContent>
        <div className='territory-cards--list'>
          {cards}
        </div>
      </DialogContent>
      <CustomButton id='territory-cards--button' onClick={tradeCards} label='Exchange Cards' disabled={!areSelectedCardsExchangeable()} />
    </Dialog>
  );
}

function Card(props: { card: TerritoryCard, index: number, isSelected: boolean, onClick: any }): JSX.Element {
  let className = 'territory-cards--card';

  if (props.isSelected) {
    className += ' selected';
  }

  function getImageSource(symbol: Symbol): string {
    if (symbol === Symbol.ARCHER) {
      return archer;
    } else if (symbol === Symbol.CAVALRY) {
      return nazgul;
    } else if (symbol === Symbol.EAGLE) {
      return eagle;
    } else {
      return wildCard;
    }
  }

  const symbol = props.card.getSymbolValue();
  return (
    <div className={className} key={props.index} onClick={props.onClick}>
      <div className='territory-card--card_symbol'>
        <img className={`symbol-${symbol}`} src={getImageSource(symbol)} alt={symbol}/>
      </div>
      <p>
        {symbol}
      </p>
    </div>
  );
}