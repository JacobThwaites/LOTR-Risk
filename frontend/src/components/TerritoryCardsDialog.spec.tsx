import '@testing-library/jest-dom'
import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react'
import TerritoryCardsDialog from './TerritoryCardsDialog';
import { TerritoryCard } from '../gameLogic/Models/TerritoryCard';


it('renders TerritoryCardsDialog correctly', () => {
    const onClose = () => {};
    const sendTradeTerritoryCardsMessage = () => {};
    const cards: TerritoryCard[] = [];

    const { container } = render(<TerritoryCardsDialog onClose={onClose} cards={cards} sendTradeTerritoryCardsMessage={sendTradeTerritoryCardsMessage}/>);
    // const component = renderer.create(
    //     <TerritoryCardsDialog onClose={onClose} cards={cards} sendTradeTerritoryCardsMessage={sendTradeTerritoryCardsMessage}/>
    // );
    console.log(container);
    
    // let tree = component.toJSON();
    expect(container).toMatchSnapshot();
});

