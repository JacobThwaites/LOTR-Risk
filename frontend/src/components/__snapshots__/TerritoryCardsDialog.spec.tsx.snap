// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders TerritoryCardsDialog correctly 1`] = `
<div
  aria-hidden="true"
/>
`;
